2-completed-app/webapp/helper.py

```diff
@@ -0,0 +1,115 @@
+        partition_key = request.form.get("StationName")
+        row_key = request.form.get("ObservationDate")
+        return self.table_client.delete_entity(partition_key, row_key)
+
+    def insert_entity(self):
+        entity = self.deserialize()
+        return self.table_client.create_entity(entity)
+
+    def upsert_entity(self):
+        entity = self.deserialize()
+        return self.table_client.upsert_entity(entity)
+
+    def update_entity(self):
+        entity = self.update_deserialize()
+        return self.table_client.update_entity(entity)
+
+    def submit_transaction(self, operations):
+        return self.table_client.submit_transaction(operations)
+
+    def insert_sample_data(self):
+        city = request.form.get('City')
+        unit = request.form.get("Units")
+        sample_data_path = os.path.join(os.getenv("project_root_path"), "webapp/data/sample_data.json")
+        weather_data = json.load(open(sample_data_path))
+        batch_insert_entity_operations = []
+        for w in weather_data:
+            if w["StationName"] == city:
+                entity = {
+                    "PartitionKey": w["StationName"],
+                    "RowKey": "{} {}".format(w["ObservationDate"], w["ObservationTime"]),
+                    "Temperature": round((w["Temperature"] - 32) * 5 / 9, 2) if unit == "US" else w["Temperature"],
+                    "Humidity": w["Humidity"],
+                    "Barometer": round(w["Barometer"] * 33.864, 2) if unit == "US" else w["Barometer"],
+                    "WindDirection": w["WindDirection"],
+                    "WindSpeed": round(w["WindSpeed"] * 1.609, 2) if unit == "US" else w["WindSpeed"],
+                    "Precipitation": round(w["Precipitation"] * 25.4, 2) if unit == "US" else w["Precipitation"]
+                }
+                batch_insert_entity_operations.append(("upsert", entity))
+        self.submit_transaction(batch_insert_entity_operations)
+
+    @staticmethod
+    def deserialize():
+        params = {key: request.form.get(key) for key in request.form.keys()}
+        params["PartitionKey"] = params.pop("StationName")
+        params["RowKey"] = "{} {}".format(params.pop("ObservationDate"), params.pop("ObservationTime"))
+        return params
+
+    @staticmethod
+    def update_deserialize():
+        params = {key: request.form.get(key) for key in request.form.keys()}
+        params["PartitionKey"] = params.pop("StationName")
+        params["RowKey"] = params.pop("ObservationDate")
+        return params
+
+    @staticmethod
+    def serializer(data):
+        field_list = []
+        for item in data:
+            for key in item.keys():
+                if key not in field_list and key not in ("PartitionKey", "RowKey"):
+                    field_list.append(key)
+        entity_list = []
+        for item in data:
+            entity = {
+                "StationName": item["PartitionKey"],
+                "ObservationDate": item["RowKey"],
```

Comment:
Does this need to be separated into "ObservationDate" and "ObservationTime"?

Reply:
@annatisch Since the front end passes `ObservationDate` and `ObservationTime` parameters separately, the input is passed separately. Besides, we wrote the repo reference to logic of [msdocs-azure-data-tables-sdk-dotnet](https://github.com/Azure-Samples/msdocs-azure-data-tables-sdk-dotnet), and the dotnet front end also passes both parameters separately. Therefore, I think there is no need to change it at present.